File: divedave-web/src/components/scene/DavePlayer.js
import {
  ANGULAR_DRAG,
  BOOST_OK_MS,
  DAVE_SPAWN_Y,
  DAVE_SPEED,
  DRAG,
  JUMP_VELOCITY,
  MAX_BOOST,
  MAX_SPIN_VELOCITY,
  MIN_SPIN_VELOCITY,
  WIDTH,
} from "../../util/Constants.js";
import { diff } from "../../util/Utilities.js";
import { BoostTiming, classifyBoostTiming } from "./DiveScorer.js";

const EARLY_TAP_WINDOW_MS = BOOST_OK_MS;
// Per-frame duration of the grounded direction-change pivot (5 frames).
const TURN_FRAME_MS = 50;
// How long the falling frame is held when tucking straight from the dive pose.
const TUCK_FALL_MS = 80;

export const DaveState = Object.freeze({
  Grounded: "grounded",
  Launching: "launching",
  Airborne: "airborne",
  Diving: "diving",
  Splashed: "splashed",
});

const ALLOWED_TRANSITIONS = {
  [DaveState.Grounded]: new Set([
    DaveState.Launching,
    DaveState.Airborne,
    DaveState.Splashed,
  ]),
  [DaveState.Launching]: new Set([DaveState.Airborne]),
  [DaveState.Airborne]: new Set([
    DaveState.Diving,
    DaveState.Grounded,
    DaveState.Splashed,
  ]),
  [DaveState.Diving]: new Set([DaveState.Splashed]),
  [DaveState.Splashed]: new Set(),
};

export class DavePlayer {
  constructor(scene, springboard) {
    this.scene = scene;
    this.springboard = springboard;

    const sprite = scene.physics.add
      .sprite(springboard.x - springboard.width / 4, DAVE_SPAWN_Y, "dave")
      .setDepth(12);
    sprite.setOrigin(0.5, 0.5);
    sprite.body.setSize(64, 256);
    sprite.body.setAllowGravity(true);
    sprite.speed = DAVE_SPEED;
    sprite.body.setAngularDrag(ANGULAR_DRAG);
    sprite.body.setAllowDrag(true);
    this.sprite = sprite;

    this.state = DaveState.Airborne;
    this.landedAt = null;
    this.jumpReleasedAt = null;
    this.boost = 0;
    this.currentVelocity = MIN_SPIN_VELOCITY;
    this.tucked = false;
    this.tuckCount = 0;
    // Facing + turn-on-reversal (grounded only). facing: 1 = right, -1 = left.
    this.facing = 1;
    this.prevDesired = 0;
    this.turning = false;
    this.turnFrom = 1;
    this.turnTo = 1;
    this.turnViaBack = false;
    this.turnStart = 0;
    // Tuck transition: slip the falling frame in when tucking from the dive pose.
    this.wasTucked = false;
    this.tuckTransitioning = false;
    this.tuckTransitionStart = 0;
    this.lastFrame = 11;

    sprite.on(Phaser.Animations.Events.ANIMATION_COMPLETE, (anim) => {
      if (!anim || anim.key !== "jump") return;
      const timing = this.calculateBoost();
      this.landedAt = null;
      sprite.setVelocityY(-JUMP_VELOCITY - this.boost);
      this.transition(DaveState.Airborne);
      if (typeof this.scene.onJumpBoostApplied === "function") {
        this.scene.onJumpBoostApplied(timing);
      }
    });
  }

  transition(next) {
    if (next === this.state) return true;
    const allowed = ALLOWED_TRANSITIONS[this.state];
    if (!allowed || !allowed.has(next)) {
      console.warn(
        `[DavePlayer] rejected transition ${this.state} → ${next}`
      );
      return false;
    }
    const prev = this.state;
    this.state = next;
    this.didEnter(next, prev);
    return true;
  }

  didEnter(next /*, prev */) {
    if (next === DaveState.Diving) {
      this.sprite.body.checkCollision.none = true;
      return;
    }
    if (next === DaveState.Splashed) {
      return;
    }
    if (next === DaveState.Grounded) {
      this.turning = false;
      this.prevDesired = 0;
      this.tucked = false;
      this.tuckCount = 0;
      this.currentVelocity = MIN_SPIN_VELOCITY;
      if (typeof this.scene.resetDiveAttempt === "function") {
        this.scene.resetDiveAttempt();
      }
      // jumpReleasedAt is intentionally NOT cleared — calculateBoost()
      // reads it to score the next jump's quickness.
      if (
        this.jumpReleasedAt &&
        Date.now() - this.jumpReleasedAt < EARLY_TAP_WINDOW_MS
      ) {
        this.tryJump();
      }
    }
  }

  tryJump() {
    if (this.state !== DaveState.Grounded) return false;
    if (this.sprite.anims.getName() === "jump") return false;
    if (!this.transition(DaveState.Launching)) return false;
    this.sprite.setFlipX(false);
    this.springboard.anims.play("flex", true);
    this.sprite.anims.play("jump", true);
    return true;
  }

  applyTuck() {
    if (this.state !== DaveState.Airborne && this.state !== DaveState.Diving) {
      return;
    }
    if (!this.tucked) {
      this.tucked = true;
      this.tuckCount++;
      if (this.state === DaveState.Airborne) {
        this.transition(DaveState.Diving);
      }
    }
    if (this.currentVelocity < MAX_SPIN_VELOCITY - 200) {
      this.currentVelocity += 5;
    } else if (this.currentVelocity < MAX_SPIN_VELOCITY) {
      this.currentVelocity += 1;
    }
    this.sprite.body.setAngularVelocity(this.currentVelocity);
  }

  releaseTuck() {
    this.tucked = false;
    this.currentVelocity = MIN_SPIN_VELOCITY;
  }

  noteJumpReleased() {
    this.jumpReleasedAt = Date.now();
  }

  noteBoardLanded() {
    if (!this.landedAt) this.landedAt = Date.now();
  }

  applyDamping() {
    this.sprite.body.velocity.x *= DRAG;
  }

  calculateBoost() {
    const quickness = diff(this.landedAt, this.jumpReleasedAt);
    const timing = classifyBoostTiming(quickness);
    switch (timing) {
      case BoostTiming.Perfect: this.boost = MAX_BOOST; break;
      case BoostTiming.Good:    this.boost = MAX_BOOST / 2; break;
      default:                  this.boost = 0;
    }

    const daveBoardDist =
      this.sprite.x - (this.springboard.x - this.springboard.width / 2);
    if (daveBoardDist > 0) {
      let newRatio = daveBoardDist / this.springboard.width;
      if (newRatio > 1) newRatio = 1;
      this.boost = this.boost * newRatio;
    }

    return timing;
  }

  isAboveBoard() {
    const d = this.sprite;
    const b = this.springboard;
    return (
      d.x + d.width / 4 > 0 &&
      d.x - d.width / 4 < b.x + b.width / 2 - 10 &&
      d.y + d.height / 2 < b.y - b.height / 2 + 1
    );
  }

  isTouchingBoard() {
    return (
      this.sprite.y + this.sprite.height / 2 >=
      this.springboard.y - this.springboard.height / 2
    );
  }

  isCleanLanding() {
    const body = this.sprite.body;
    if (!body) return false;
    const notLaunching = body.velocity.y > -50;
    const lowSpin = Math.abs(body.angularVelocity) < 30;
    return notLaunching && lowSpin;
  }

  isTucked() {
    if (this.tucked) this.sprite.setFrame(13);
    return this.sprite.frame.name === 13 || this.tucked;
  }

  updateFrame() {
    const dave = this.sprite;
    switch (this.state) {
      case DaveState.Launching:
      case DaveState.Splashed:
        return;

      case DaveState.Grounded: {
        if (dave.angle !== 0) dave.setAngle(0);
        if (this.turning) {
          this.advanceTurn();
          return;
        }
        const thr = DAVE_SPEED / 10;
        const vx = dave.body.velocity.x;
        const desired = vx > thr ? 1 : vx < -thr ? -1 : 0;
        if (desired === 0) {
          dave.setFlipX(this.facing < 0);
          if (dave.anims.getName() !== "idle") dave.anims.play("idle", true);
          this.prevDesired = 0;
        } else if (desired === this.facing) {
          dave.setFlipX(this.facing < 0);
          dave.anims.play("walk", true);
          this.prevDesired = desired;
        } else if (this.prevDesired === 0) {
          // From idle/landing: snap to the new direction, no turn.
          this.facing = desired;
          dave.setFlipX(this.facing < 0);
          dave.anims.play("walk", true);
          this.prevDesired = desired;
        } else {
          // Reversed mid-walk: pivot through the turn frames, then resume walk.
          this.startTurn(this.facing, desired);
          this.prevDesired = desired;
        }
        return;
      }

      case DaveState.Airborne:
        // Frames face right; mirror for leftward motion. Ascending = jump-up
        // (11), descending = falling-down (12).
        dave.setFlipX(dave.body.velocity.x < 0);
        this.lastFrame = dave.body.velocity.y < 0 ? 11 : 12;
        dave.setFrame(this.lastFrame);
        return;

      case DaveState.Diving: {
        // Dive pose never mirrors — always the original (right-facing) frame.
        dave.setFlipX(false);
        if (this.tucked) {
          // Tucking straight from the dive (inverted) frame slips the falling
          // frame in first; from the falling/upright frame it's skipped.
          if (!this.wasTucked && this.lastFrame === 14) {
            this.tuckTransitioning = true;
            this.tuckTransitionStart = this.scene.time.now;
          }
          if (
            this.tuckTransitioning &&
            this.scene.time.now - this.tuckTransitionStart < TUCK_FALL_MS
          ) {
            this.lastFrame = 12;
          } else {
            this.tuckTransitioning = false;
            this.lastFrame = 13;
          }
        } else {
          this.tuckTransitioning = false;
          // Falling pose while upright; dive pose once rotated upside down.
          const upright = dave.angle >= -90 && dave.angle <= 90;
          this.lastFrame = upright ? 12 : 14;
        }
        dave.setFrame(this.lastFrame);
        this.wasTucked = this.tucked;
        return;
      }
    }
  }

  startTurn(from, to) {
    this.turning = true;
    this.turnFrom = from;
    this.turnTo = to;
    this.turnViaBack = Math.random() < 0.5;
    this.turnStart = this.scene.time.now;
    this.sprite.anims.stop();
    this.advanceTurn();
  }

  advanceTurn() {
    const mid = this.turnViaBack ? 3 : 1;
    const pivot = this.turnViaBack ? 4 : 0;
    const frames = [2, mid, pivot, mid, 2];
    const dirs = [
      this.turnFrom,
      this.turnFrom,
      this.turnFrom,
      this.turnTo,
      this.turnTo,
    ];
    const step = Math.floor((this.scene.time.now - this.turnStart) / TURN_FRAME_MS);
    if (step >= frames.length) {
      this.turning = false;
      this.facing = this.turnTo;
      this.sprite.setFlipX(this.facing < 0);
      this.sprite.setFrame(2);
      return;
    }
    this.sprite.setFlipX(dirs[step] < 0);
    this.sprite.setFrame(frames[step]);
  }
}

File: divedave-ios/divedave Shared/Components/Scene/DavePlayer.swift
import SpriteKit
import os

private let logger = Logger(subsystem: "com.drawvid.divedave", category: "dave-state")

enum DaveState {
    case grounded
    case launching
    case airborne
    case diving
    case splashed
}

@MainActor
final class DavePlayer {
    let dave: AnimatedSprite

    private(set) var state: DaveState = .airborne

    var boost: CGFloat = 0
    var landedAt: CFTimeInterval = 0

    // Facing + turn-on-reversal (grounded only). facing: 1 = right, -1 = left.
    private var facing: CGFloat = 1
    private var prevDesired: CGFloat = 0
    private var turning = false
    private var turnFrom: CGFloat = 1
    private var turnTo: CGFloat = 1
    private var turnViaBack = false
    private var turnStart: TimeInterval = 0
    private let turnFrameDuration: TimeInterval = 0.05

    // Tuck transition: slip the falling frame in when tucking from the dive pose.
    private var wasTucked = false
    private var tuckTransitioning = false
    private var tuckTransitionStart: TimeInterval = 0
    private var lastFrame = 11
    private let tuckFallDuration: TimeInterval = 0.08

    private static let allowedTransitions: [DaveState: Set<DaveState>] = [
        .grounded:  [.launching, .airborne],
        .launching: [.airborne],
        .airborne:  [.diving, .grounded, .splashed],
        .diving:    [.splashed],
        .splashed:  []
    ]

    init(scene: SKScene, springboard: AnimatedSprite) {
        let d = AnimatedSprite(spritesheetName: "divedave-spritesheet-extruded",
                               frameWidth: Game.defaultDaveHeight,
                               frameHeight: Game.defaultDaveHeight,
                               margin: 1,
                               spacing: 2)
        d.position = CGPoint(x: springboard.frame.minX + springboard.frame.width / 4,
                             y: springboard.position.y + 100)
        d.zPosition = 5

        d.physicsBody = SKPhysicsBody(rectangleOf: CGSize(width: d.size.width / 4, height: d.size.height))
        d.physicsBody?.isDynamic = true
        d.physicsBody?.mass = Game.daveMass
        d.physicsBody?.affectedByGravity = true
        d.physicsBody?.restitution = 0.0
        d.physicsBody?.friction = 0.0

        d.physicsBody?.categoryBitMask = PhysicsCategory.dave.rawValue
        d.physicsBody?.contactTestBitMask = PhysicsCategory.springboard.rawValue
        d.physicsBody?.collisionBitMask = PhysicsCategory.springboard.rawValue

        scene.addChild(d)

        d.defineAnimation(name: "idle", frameIndices: [15, 15, 15, 15, 15, 16, 17, 18], timePerFrame: 0.125)
        d.defineAnimation(name: "walk", frameIndices: [5, 6, 7, 8], timePerFrame: 0.125)
        d.defineAnimation(name: "jump", frameIndices: [10, 10, 11], timePerFrame: 0.1, repeatForever: false)

        self.dave = d
    }

    @discardableResult
    func transition(to next: DaveState) -> Bool {
        if next == state { return true }
        guard Self.allowedTransitions[state]?.contains(next) == true else {
            logger.debug("rejected dave transition: \(String(describing: self.state)) -> \(String(describing: next))")
            return false
        }
        logger.debug("dave transition: \(String(describing: self.state)) -> \(String(describing: next))")
        state = next
        didEnter(next)
        return true
    }

    private func didEnter(_ state: DaveState) {
        switch state {
        case .diving:
            // Drop board collision once committed to a dive — Dave passes through.
            dave.physicsBody?.collisionBitMask = 0
            dave.physicsBody?.contactTestBitMask = 0
        case .grounded:
            turning = false
            prevDesired = 0
        case .launching, .airborne, .splashed:
            break
        }
    }

    func isCleanLanding() -> Bool {
        guard let body = dave.physicsBody else { return false }
        // dy tolerance instead of <= 0: SpriteKit's collision resolution leaves
        // tiny floating-point noise (~4e-12) at the instant didBegin fires.
        // A strict check would reject real landings as "ascending".
        let notLaunching = body.velocity.dy < 50
        // 30°/s in radians — matches web's Phaser angularVelocity gate (deg/s).
        let lowSpin = abs(body.angularVelocity) < 30.0 * .pi / 180.0
        return notLaunching && lowSpin
    }

    func jump(
        springboard: AnimatedSprite,
        onJumpStarted: (() -> Void)? = nil,
        onJumpCompleted: ((BoostTiming) -> Void)? = nil
    ) {
        guard state == .grounded else { return }
        guard transition(to: .launching) else { return }

        onJumpStarted?()

        setFacing(left: false)
        dave.playAnimation(name: "jump") { [weak self] in
            guard let self = self else { return }
            let timing = self.calculateBoost(springboard: springboard)
            self.playBoardBounce(springboard: springboard, timing: timing)
            self.landedAt = 0
            self.dave.physicsBody?.velocity.dy = Game.jumpVelocity + self.boost
            self.transition(to: .airborne)
            onJumpCompleted?(timing)
        }
    }

    // Flex the board to this jump's timing frame (1 = OK, 2 = GOOD, 3 = PERFECT);
    // a non-notable (miss) bounce leaves it at rest. Mirrors divedave-web.
    private func playBoardBounce(springboard: AnimatedSprite, timing: BoostTiming) {
        let bounce: String
        switch timing {
        case .perfect: bounce = "bouncePerfect"
        case .good:    bounce = "bounceGood"
        case .ok:      bounce = "bounceOk"
        case .miss:    return
        }
        springboard.playAnimation(name: bounce) {
            springboard.clearCurrentAnimation()
        }
    }

    @discardableResult
    private func calculateBoost(springboard: AnimatedSprite) -> BoostTiming {
        let quickness = abs(Self.msBetween(landedAt, GameState.shared.jumpReleasedAt))
        let timing = DiveScorer.classifyBoostTiming(quicknessMs: quickness)
        switch timing {
        case .perfect:   boost = Game.maxBoost
        case .good:      boost = Game.maxBoost / 2
        case .ok, .miss: boost = 0
        }

        let daveBoardDist = dave.position.x - (springboard.position.x - springboard.size.width / 2)
        if daveBoardDist > 0 {
            var newRatio = daveBoardDist / springboard.size.width
            newRatio = min(newRatio, 1)
            boost *= newRatio
        }

        return timing
    }

    private static func msBetween(_ start: CFTimeInterval, _ end: CFTimeInterval) -> Double {
        guard start > 0, end > 0 else { return .greatestFiniteMagnitude }
        return (end - start) * 1000
    }

    func applyDamping() {
        guard let body = dave.physicsBody else { return }
        let newVelocityX = body.velocity.dx * Game.drag
        body.velocity = CGVector(dx: newVelocityX, dy: body.velocity.dy)
    }

    // Frames face right; a negative xScale mirrors Dave for leftward motion.
    // The physics body is a centered rectangle, so mirroring leaves it unchanged.
    private func setFacing(left: Bool) {
        dave.xScale = left ? -abs(dave.xScale) : abs(dave.xScale)
    }

    func updateFrame(tucked: Bool, currentTime: TimeInterval) {
        switch state {
        case .launching, .splashed:
            return

        case .grounded:
            if dave.zRotation != 0 { dave.zRotation = 0 }
            if turning {
                advanceTurn(currentTime: currentTime)
                return
            }
            let thr = Game.daveSpeed / 10
            let dx = dave.physicsBody?.velocity.dx ?? 0
            let desired: CGFloat = dx > thr ? 1 : (dx < -thr ? -1 : 0)
            if desired == 0 {
                setFacing(left: facing < 0)
                dave.playAnimation(name: "idle")
                prevDesired = 0
            } else if desired == facing {
                setFacing(left: facing < 0)
                dave.playAnimation(name: "walk")
                prevDesired = desired
            } else if prevDesired == 0 {
                // From idle/landing: snap to the new direction, no turn.
                facing = desired
                setFacing(left: facing < 0)
                dave.playAnimation(name: "walk")
                prevDesired = desired
            } else {
                // Reversed mid-walk: pivot through the turn frames, then walk.
                startTurn(from: facing, to: desired, currentTime: currentTime)
                prevDesired = desired
            }

        case .airborne:
            dave.stopAnimation()
            let dx = dave.physicsBody?.velocity.dx ?? 0
            let dy = dave.physicsBody?.velocity.dy ?? 0
            setFacing(left: dx < 0)
            // dy > 0 ascending = jump-up (11), else falling-down (12)
            lastFrame = dy > 0 ? 11 : 12
            dave.texture = dave.frames[lastFrame]

        case .diving:
            dave.stopAnimation()
            // Dive pose never mirrors — always the original (right-facing) frame.
            setFacing(left: false)
            if tucked {
                // Tucking straight from the dive (inverted) frame slips the
                // falling frame in first; from the falling/upright frame it's skipped.
                if !wasTucked && lastFrame == 14 {
                    tuckTransitioning = true
                    tuckTransitionStart = currentTime
                }
                if tuckTransitioning && currentTime - tuckTransitionStart < tuckFallDuration {
                    lastFrame = 12
                } else {
                    tuckTransitioning = false
                    lastFrame = 13
                }
            } else {
                tuckTransitioning = false
                // Falling pose while upright; dive pose once rotated upside down.
                // Normalize accumulated spin to [-pi, pi] before the test.
                let twoPi = 2 * CGFloat.pi
                var r = dave.zRotation.truncatingRemainder(dividingBy: twoPi)
                if r > .pi { r -= twoPi } else if r < -.pi { r += twoPi }
                let upright = r >= -CGFloat.pi / 2 && r <= CGFloat.pi / 2
                lastFrame = upright ? 12 : 14
            }
            dave.texture = dave.frames[lastFrame]
            wasTucked = tucked
        }
    }

    private func startTurn(from: CGFloat, to: CGFloat, currentTime: TimeInterval) {
        turning = true
        turnFrom = from
        turnTo = to
        turnViaBack = Bool.random()
        turnStart = currentTime
        dave.stopAnimation()
        advanceTurn(currentTime: currentTime)
    }

    private func advanceTurn(currentTime: TimeInterval) {
        let mid = turnViaBack ? 3 : 1
        let pivot = turnViaBack ? 4 : 0
        let frames = [2, mid, pivot, mid, 2]
        let dirs = [turnFrom, turnFrom, turnFrom, turnTo, turnTo]
        let step = Int((currentTime - turnStart) / turnFrameDuration)
        if step >= frames.count {
            turning = false
            facing = turnTo
            setFacing(left: facing < 0)
            dave.texture = dave.frames[2]
            return
        }
        setFacing(left: dirs[step] < 0)
        dave.texture = dave.frames[frames[step]]
    }
}
